client/allocrunner/task_hook_coordinator.go

```diff
@@ -80,31 +108,53 @@ func (c *taskHookCoordinator) startConditionForTask(task *structs.Task) <-chan s
 
 // This is not thread safe! This must only be called from one thread per alloc runner.
 func (c *taskHookCoordinator) taskStateUpdated(states map[string]*structs.TaskState) {
-	if c.mainTaskCtx.Err() != nil {
+	if c.mainTaskCtx.Err() != nil && c.poststopTaskCtx.Err() != nil {
 		// nothing to do here
 		return
 	}
 
-	for task := range c.prestartSidecar {
+	for task := range c.prestartSidecarTasks {
 		st := states[task]
 		if st == nil || st.StartedAt.IsZero() {
 			continue
 		}
 
-		delete(c.prestartSidecar, task)
+		delete(c.prestartSidecarTasks, task)
 	}
 
-	for task := range c.prestartEphemeral {
+	for task := range c.prestartEphemeralTasks {
 		st := states[task]
 		if st == nil || !st.Successful() {
 			continue
 		}
 
-		delete(c.prestartEphemeral, task)
+		delete(c.prestartEphemeralTasks, task)
+	}
+
+	for task := range c.mainTasks {
+		st := states[task]
+		if st == nil || !st.Successful() {
+			continue
+		}
+
+		delete(c.mainTasks, task)
+	}
+
+	for task := range c.poststopTasks {
+		st := states[task]
+		if st == nil || !st.Successful() {
+			continue
+		}
+
+		delete(c.poststopTasks, task)
 	}
```

Comment:
Do we need to track post stop tasks?  We don't depend on their state currently, as nothing is blocked by them.